languages/rust/exercises/concept/enums/.docs/instructions.md

```diff
@@ -0,0 +1,54 @@
+# Instructions
+
+In this exercise you'll be processing log lines.
+
+Each log line is a string formatted as follows: `"[<LEVEL>]: <MESSAGE>"`.
+
+There are three different log levels:
+
+- `INFO`
+- `WARNING`
+- `ERROR`
+
+You have three tasks.
+
+### 1. Parse log level
+
+Define a `LogLevel` enum that has three elements:
+
+- `Info`
+- `Warning`
+- `Error`
+
+Next, write a function to parse the log level of a log line:
```

Comment:
By far the simplest approach to this particular problem uses regular expressions. That's not a problem when students are solving this exercise on their own machines; they can just import the regex crate. However, that doesn't work when  solving exercises online: the online editor doesn't support editing `Cargo.toml` or downloading additional crates. 

Perhaps we should invert the sense of this exercise so students are emitting structured log messages, instead of parsing text log lines, to reduce that problem?

Reply:
> By far the simplest approach to this particular problem uses regular expressions. 

I agree. `&str` methods work but feel verbose.

> Perhaps we should invert the sense of this exercise so students are emitting structured log messages, instead of parsing text log lines, to reduce that problem?

Good idea! Would this be emitting a struct with a `type` field set to the `enum`? Or could you post a snippet demonstrating what you were thinking?

Given #663, it seems like we could simplify as well to fit Rust's idioms better.

I am unclear on how far we should deviate from the csharp concept exercises; I'm leaning toward copying them more verbosely at first to quickly finish a few so we can refine the concepts further and write the concept exercise implementation guide.

Reply:
Well, we can't do real structured logs if the students haven't done at least declarative macros, so let's call these semistructured. The idea is that we provide the `LogLevel` enum (omitting `Unknown`) and then require students to implement methods like this:

```rust
pub fn log(level: LogLevel, message: &str) -> String { unimplemented!() }
pub fn info(message: &str) -> String { unimplemented!() }
pub fn warn(message: &str) -> String { unimplemented!() }
pub fn error(message: &str) -> String { unimplemented!() }
```

TBH, I'm not a huge fan of the short-log portion of this exercise: while it teaches that some enums can be represented by integer primitives, actually converting between enums and integers in Rust is an antipattern in every case except serialization. I'd just skip that.